### PR TITLE
feat(cli): update CLI to use new deploy endpoint

### PR DESCRIPTION
**Pull request by ricokahler**
NOTE: This PR cannot be merged until the newer deployment service is in prod.

### Description

This PR introduces several key changes to the Sanity CLI:
1. **New Deploy Endpoint**: The CLI now uses a new endpoint for deploying Sanity Studios, aligning with the updated deployment infrastructure.
2. **Multiple `studioHost` Support**: Projects can now have multiple `studioHost` configurations, allowing for multiple deployments under different hostnames. This is facilitated by the new `studioHost` option in the CLI configuration.
3. **Helper Functions Refactor**: Functions such as `getOrCreateUserApplication`, `createDeployment`, `deleteUserApplication`, and others have been refactored and relocated to a separate `helpers.ts` file for better modularity and testability.
4. (edit) **Use of `form-data` lib + ~~native `fetch` API~~ `client.request`**: I added an additional commit that utilizes the `form-data` library to re-enabling streaming of the tarball to the server instead of buffering all of it into memory.

These changes are part of the larger multi-deploy studio project, aiming to enhance the deployment capabilities and flexibility of the Sanity CLI.

### What to review

1. **Configuration Changes**: Review the addition of the `studioHost` configuration option in `CliConfig`.
2. **Deployment Flow**: Ensure the new deploy and undeploy flows work correctly, including handling multiple `studioHost` values.
3. **Test Coverage**: Check the new and updated tests for `deployAction`, `undeployAction`, and the functions in `helpers.ts` for adequate coverage and correctness.
5. **Code Quality**: Review the refactoring in `deployAction.ts` and `undeployAction.ts` for adherence to best practices and consistency with the existing codebase.

### Testing

Extensive unit tests have been added and updated to cover:
- The new deploy and undeploy workflows, including handling of multiple `studioHost` configurations.
- The new helper functions for managing user applications and deployments.
- Edge cases, such as missing directories and error handling during the deployment process.

These tests ensure that the changes are robust and handle various scenarios correctly. Please refer to the new test files and sections within existing tests for detailed coverage.

Note, the new service is not in prod yet so these changes have not been tested e2e yet.

### Notes for release

TBD

### CODE REVIEW

File: packages/sanity/src/_internal/cli/actions/deploy/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+  version,
+}: CreateDeploymentOptions): Promise<void> {
+  const formData = new FormData()
+  formData.append('isAutoUpdating', isAutoUpdating.toString())
+  formData.append('version', version)
+  formData.append('tarball', tarball, {contentType: 'application/gzip', filename: 'app.tar.gz'})
+
+  await client.request({
+    uri: `/user-applications/${applicationId}/deployments`,
+    method: 'POST',
```
Comment (binoy14) [inline packages/sanity/src/_internal/cli/actions/deploy/helpers.ts:154]: ❓ 🤔 would this need `headers: { ...formData.getHeaders() }` to get the correct multipart boundary?

**Comment by binoy14** [comment]
* **#7244** <a href="https://app.graphite.dev/github/pr/sanity-io/sanity/7244?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @ricokahler and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/sanity/src/_internal/cli/actions/deploy/undeployAction.ts [hunk 2]
```diff
@@ -1,23 +1,32 @@
...
+  const userApplication = await getUserApplication({
+    client,
+    appHost: cliConfig && 'studioHost' in cliConfig ? cliConfig.studioHost : undefined,
+  })
+
   spinner.succeed()
 
-  if (!studioHost) {
+  if (!userApplication) {
     output.print('Your project has not been assigned a studio hostname.')
```
Comment (cngonzalez) [inline packages/sanity/src/_internal/cli/actions/deploy/undeployAction.ts:30]: I found this output a little bit confusing -- it could also just mean that they DO have the hostname but DON'T have a default and didn't put it in CLI (which will happen to every one starting a new project from here on out.)

```suggestion
    output.print('Your project has not been assigned a studio hostname,')
     output.print('or you do not have studioHost set in in sanity.cli.js or sanity.cli.ts.')
```

**Review by cngonzalez — APPROVED** [review]
One tiny comment for UX / DX, but otherwise logic makes sense to me and I was not able to break in manual testing. Thanks for adding all the tests!

File: packages/sanity/src/_internal/cli/actions/deploy/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+ *           +-----------+---------------+
+ *                       |
+ *                +------v------+
+ *                |             |
+ *                v             v
+ *   +--------------------+  +------------------------+
+ *   | Show list and      |  | Prompt for hostname    |
+ *   | prompt selection   |  | and create new app     |
+ *   +--------------------+  +------------------------+
+ */
```
Comment (ricokahler) [inline packages/sanity/src/_internal/cli/actions/deploy/helpers.ts:137]: very nice internal docs!

File: packages/sanity/src/_internal/cli/actions/deploy/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+    // if the user selected an existing app, return it
+    if (selected !== 'new') {
+      return userApplications.find((app) => app.appHost === selected)!
+    }
+  }
+
+  // otherwise, prompt the user for a hostname
+  output.print('Your project has not been assigned a studio hostname.')
+  output.print('To deploy your Sanity Studio to our hosted Sanity.Studio service,')
```
Comment (cngonzalez) [inline packages/sanity/src/_internal/cli/actions/deploy/helpers.ts:180]: maybe no caps here?
```suggestion
  output.print('To deploy your Sanity Studio to our hosted sanity.studio service,')
```

File: packages/sanity/src/_internal/cli/actions/deploy/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+      urlType: 'internal',
+    })
+    spinner.succeed()
+
+    return response
+  } catch (e) {
+    spinner.fail()
+    // if the name is taken, it should return a 409 so we relay to the user
+    if ([402, 409].includes(e?.statusCode)) {
+      throw new Error(e?.response?.body?.message) || 'Bad request' // just in case
```
Comment (ricokahler) [inline packages/sanity/src/_internal/cli/actions/deploy/helpers.ts:257]: did you mean?

```suggestion
      throw new Error(e?.response?.body?.message || 'Bad request') // just in case
```

**Review by ricokahler — COMMENTED** [review]
the only thing i caught was maybe throwing the string

File: packages/sanity/src/_internal/cli/actions/deploy/helpers.ts (hunk 5 above)
Comment (binoy14) [reply packages/sanity/src/_internal/cli/actions/deploy/helpers.ts:257]: Good catch! thanks!

**Review by cngonzalez — APPROVED** [review]
